### PR TITLE
oomd: fix potential race between killing cgroups and cleaning up cgroups

### PR DESCRIPTION
**Pull request by anitazha**
Fixes: #22030

### CODE REVIEW

**Comment by mbiebl** [comment]
Should be Fixes: #22030 

I'll close the (now outdated) PR #22029

**Comment by mbiebl** [comment]
I ran 10 iterations of the test, all of them passed.
Thanks a lot @anitazha !

**Comment by anitazha** [comment]
Updated the issue number and switched ret value checking to use `IN_SET()`

File: src/oom/oomd-util.c [hunk 1]
```diff
@@ -196,9 +196,19 @@ int oomd_cgroup_kill(const char *path, bool recurse, bool dry_run) {
...
         else
                 r = cg_kill(SYSTEMD_CGROUP_CONTROLLER, path, SIGKILL, CGROUP_IGNORE_SELF, pids_killed, log_kill, NULL);
-        if (r < 0)
+
+        /* The cgroup could have been cleaned up after we have sent SIGKILL to all of the processes, but before
+         * we could do one last iteration of cgroup.procs to check. Or the service unit could have exited and
+         * was removed between picking candidates and coming into this function. In either case, let's log
+         * about it let the caller decide what to do once they know how many PIDs were killed. */
+        if (IN_SET(r, -ENOENT, -ENODEV))
+                log_debug_errno(r, "Error when killing cgroup path %s: %m", path);
```
Comment (yuwata) [inline src/oom/oomd-util.c]: I'd like to the message be more informative. And, I think it is better to add `", ignoring"`.

**Review by yuwata — APPROVED** [review]
LGTM. One minor suggestion.

**Review by bluca — APPROVED** [review]

**Comment by anitazha** [comment]
Thanks! Updated with a better log message

**Pull request merged**